src/bin/vhost_user_blk.rs

```diff
@@ -71,10 +73,11 @@ struct VhostUserBlkBackend {
     disk_image_id: Vec<u8>,
     disk_nsectors: u64,
     config: virtio_blk_config,
+    rdonly: bool,
 }
 
 impl VhostUserBlkBackend {
-    pub fn new(image_path: String) -> Result<Self> {
+    pub fn new(image_path: String, rdonly: bool) -> Result<Self> {
         let raw_img: File = OpenOptions::new()
             .read(true)
             .write(true)
```

Comment:
@yangzhon isn't the point of the `readonly` feature to open the source file in `readonly` mode?
And if that's the case, then you should be modifying this block of code:
```rust
let file: File = OpenOptions::new();
        .read(true)
        .write(!readonly)
        .open(&image_path)
        .unwrap();
```

Reply:
@sboeuf ,

Yes, adding write(!readonly) into file open is one method.

We also can add this FLAG like below into avail_features   
avail_features |= 1 << VIRTIO_BLK_F_RO

Guest will get these features and we can check by ""cat /sys/block/vdb/device/features" command
0000010000000000000000000000000010000000000000000000000000000000
Bit 5 show RO attribute of this block device.

By the way, from lsblk command, we can also get if RO is 1.

Once this device is RO, the write operation for this block in guest will report "write error: Operation not permitted"  and we can do write operation.

Reply:
I think it's better to do both since we can never trust the guest for doing the right thing. Here, by opening the file as `readonly`, we add one more security to ensure the file won't be altered.

Reply:
Okay, thanks, i will add it.